File: programmers/오픈채팅방/오픈채팅방.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <sstream>
using namespace std;

vector<string> split(string target, char division) {
    vector<string> res;
    stringstream ss(target);
    string temp;
    
    while (getline(ss, temp, division)) res.push_back(temp);
    return res;
}

vector<string> solution(vector<string> record) {
    vector<string> answer;
    unordered_map<string, string> res;
    vector<pair<string, string> > result;
    
    res["Enter"] = "님이 들어왔습니다.";
    res["Leave"] = "님이 나갔습니다.";
    
    for (auto cur : record) {
        auto item = split(cur, ' ');
        if (item[0].compare("Change") == 0 || item[0].compare("Enter") == 0) res[item[1]] = item[2];
        if (item[0].compare("Leave") == 0 || item[0].compare("Enter") == 0) result.push_back({item[1], item[0]});
    }
    for (auto r : result) answer.push_back(res[r.first] + res[r.second]);
    
    return answer;
}

File: programmers/오픈채팅방/오픈채팅방.kt
import java.util.*
class Solution {
    fun solution(record: Array<String>): Array<String> {
        val userName = TreeMap<String, String>()
        val temp = mutableListOf<String>()
        record.forEach {
            val item = it.split(" ")
            if (item[0] == "Enter" || item[0] == "Change") userName[item[1]] = item[2]
            if (item[0] == "Enter" || item[0] == "Leave") temp.add("${item[1]} ${item[0]}")
        }
        var answer = mutableListOf<String>()
        temp.forEach {
            val item = it.split(" ")
            answer.add("${userName[item[0]]}${
                       if (item[1] == "Enter") "님이 들어왔습니다."
                       else "님이 나갔습니다."
                       }")
        }
        return answer.toTypedArray()
    }
}